### PR TITLE
feat(@nestjs/core): Add RouterModule to the core

### PR DESCRIPTION
**Pull request by shekohex**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no API changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #255 , #389 and #1438

## What is the new behavior?

For a long time ago (almost 2 years) I created and maintained the [nest-router](https://github.com/nestjsx/nest-router) package as a 3rd party package.
Over that time, we gained a lot of feedback and fixed a lot of issues (especially integration with other packages like @nestjs/swagger for example).

And now, we think it's time to add this package as a module into the nestjs core, and Hopefully it would be useful in a lot of use cases, for example:

1. API Versioning.
2. Arrange Routes/Modules hierarchically.

...and the list goes on.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

File: packages/core/router/router-explorer.ts [hunk 1]
```diff
@@ -89,7 +89,11 @@ export class RouterExplorer {
     prefix?: string,
   ): string {
     let path = Reflect.getMetadata(PATH_METADATA, metatype);
-    if (prefix) path = prefix + this.validateRoutePath(path);
+    if (prefix) {
+      path = prefix + this.validateRoutePath(path);
+      // Override the metadata with the new path
+      Reflect.defineMetadata(PATH_METADATA, path, metatype);
+    }
```
Comment (shekohex) [inline packages/core/router/router-explorer.ts]: I think doing this would solve all integration issues with any other packages that depend on routes

Comment (shekohex) [reply packages/core/router/router-explorer.ts]: for example the swagger integration issues and also this issue https://github.com/nestjsx/nest-router/issues/5

Comment (kamilmysliwiec) [reply packages/core/router/router-explorer.ts]: Very uncommon but still: what if I have the same controller used in 2 modules? In this case, I'd override the metadata twice.

**Comment by tonivj5** [comment]
Any news? :+1:

**Comment by kamilmysliwiec** [comment]
Let me know if there's anything we can do to help @shekohex!

**Comment by shekohex** [comment]
Hi @kamilmysliwiec 
I'm sorry, I was so busy in the past few days, but as I got some time, I'll continue working on this PR again.

The issue here with `Middlewares` as stated here https://github.com/nestjsx/nest-router/issues/5.
I made a Workaround [here](https://github.com/nestjsx/nest-router/blob/f860587da384dec45b17f8dc448b771bb7ea2074/src/router.module.ts#L42-L50) so far so good, but I think if we are gonna make this a Core Module, why do we need these workarounds.
So any suggestions to solve this issue? without doing [this](https://github.com/nestjs/nest/pull/3489#pullrequestreview-324344732) which I think it does not even [work](https://circleci.com/gh/nestjs/nest/8098)

**Comment by kamilmysliwiec** [comment]
Actually, [this](https://github.com/nestjsx/nest-router/blob/f860587da384dec45b17f8dc448b771bb7ea2074/src/router.module.ts#L42-L50) doesn't look that bad ;)

**Comment by tonivj5** [comment]
Is this PR going to be merged as part of 7.0.0 release? :rocket:

**Comment by shekohex** [comment]
Hi @kamilmysliwiec 
Any Help here with this Middleware error? I don't know how that happened

**Comment by tonivj5** [comment]
Hey, any progress on this? :eyes:

**Comment by UvealSnow** [comment]
Would love to have this feature!

**Comment by vahidvdn** [comment]
Any news?

**Comment by shekohex** [comment]
Hey all, anyone is interested to take over this PR?
As I don't have any free time to do it these days, things gone crazy this year!

**Comment by vahidvdn** [comment]
@shekohex @kamilmysliwiec  In Angular, I used to pass the custom data object where I define the route (in router module). Then in guard, I can have access to that data. (e.g. `isPublic = true`)

But here (as [described](https://docs.nestjs.com/security/authentication#enable-authentication-globally)) we should define a custom decorator and decorate the controller's method with it, then in guard, we can have access to the decorator's value (to check whether the API is decorated with that decorator or not)
I think the Angular approach is much better from the maintainability point of view which can be possible with this RouterModule.

**Comment by kamilmysliwiec** [comment]
@shekohex I've cherry-picked your changes and created a new PR https://github.com/nestjs/nest/pull/6035 let's track this there

**Comment by vahidvdn** [comment]
@kamilmysliwiec Would you please consider [this](https://github.com/nestjsx/nest-router/issues/29#issuecomment-738876167) for versioning?

**Pull request closed without merging**